cmd/PersistGrantsGovXMLDB/handler.go

```diff
@@ -0,0 +1,95 @@
+package main
+
+import (
+	"context"
+	"encoding/xml"
+	"io"
+
+	"github.com/aws/aws-lambda-go/events"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/hashicorp/go-multierror"
+	"github.com/usdigitalresponse/grants-ingest/internal/log"
+	grantsgov "github.com/usdigitalresponse/grants-ingest/pkg/grantsSchemas/grants.gov"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const (
+	MB                         = int64(1024 * 1024)
+	GRANT_OPPORTUNITY_XML_NAME = "OpportunitySynopsisDetail_1_0"
+)
+
+type opportunity grantsgov.OpportunitySynopsisDetail_1_0
+
+// handleS3Event handles events representing S3 bucket notifications of type "ObjectCreated:*"
+// for XML DB extracts saved from Grants.gov. The XML data from the source S3 object provided
+// by each event record is read from S3. Grant opportunity records are extracted from the XML
+// and uploaded to a "prepared data" destination bucket as individual S3 objects.
+// Returns an error that represents any and all errors accumulated during the invocation,
+// either while handling a source object or while processing its contents; an error may indicate
+// a partial or complete invocation failure.
+// Returns nil when all grant opportunities are successfully processed from all source records,
+// indicating complete success.
+func handleS3EventWithConfig(s3svc *s3.Client, dynamodbsvc DynamoDBUpdateItemAPI, ctx context.Context, s3Event events.S3Event) error {
+	wg := multierror.Group{}
+	for _, record := range s3Event.Records {
+		func(record events.S3EventRecord) {
+			wg.Go(func() (err error) {
+				span, ctx := tracer.StartSpanFromContext(ctx, "handle.record")
+				defer span.Finish(tracer.WithError(err))
```

Comment:
Just to stay aligned with the "there may be more than 1 opportunity" assumption that the handler code is operating under, I'd suggest moving the emission of `oportunity.failed` metrics to occur within the loop. 

Putting the conditional call to `sendMetric` in a deferred function can help keep things DRY:

```suggestion
				defer span.Finish(tracer.WithError(err))
				defer func() {
					if err != nil {
						sendMetric("opportunity.failed", 1)
					}
				}()
```

Reply:
yup, makes sense to me!